src/ripple/app/ledger/Ledger.cpp

```diff
@@ -282,6 +283,11 @@ Ledger::Ledger(Ledger const& prevLedger, NetClock::time_point closeTime)
         info_.closeTime =
             prevLedger.info_.closeTime + info_.closeTimeResolution;
     }
+
+    if (prevLedger.rules().enabled(featureNegativeUNL))
```

Comment:
I am not sure this is the best place to call `updateNegativeUNL()`, with the update happening behind the scenes. Updating the negative unl seems similar to updating the skiplist. `updateSkipList()` is called here: 
https://github.com/ripple/rippled/blob/develop/src/ripple/app/ledger/impl/BuildLedger.cpp#L60
I think `updateNegativeUNL()` should be called there too. When I use this constructor, I expect the constructed object to have the exact same state map as the previous ledger, with no differences. After that, I modify the state map based on new transations, fee voting, updating the skip list, etc. These modifications happen after the constructor. I think `updateNegativeUNL()` is a similar modification that should happen after the constructor.

Reply:
"When I use this constructor, I expect the constructed object to have the exact same state map as the previous ledger, with no differences." Is this agreed? @nbougalis @seelabs If so, we must document that. 

`updateNegativeUNL()` is private. I'd like to keep it private if possible. It must be called after the state map is copied and before applying the transactions.

Reply:
I think you should call it here: https://github.com/ripple/rippled/blob/97712107b71a8e2089d2e3fcef9ebf5362951110/src/ripple/app/ledger/impl/BuildLedger.cpp#L49
updating the negative unl is part of building the ledger, so I would call it in that function, before applying the transactions.

Reply:
For correctness, the `updateNegativeUNL()` function must be called after the state map is copied and before applying the transactions. So the place you suggested is correct, and the current code is also correct as well. In the current code, however, the function is private. I think that is a benefit, because it reduces the chance of misuse.